Upgrade Jet Factories to Multifactories

ruse-traveler opened the pull request:
### Briefly, what does this PR introduce?

This PR upgrades the existing jet factories to multifactories and also modifies the jet reconstruction algorithm such that it returns a PODIO collection of reconstructed particles rather than a standard vector.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ X] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [X ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?

No.

### Does this PR change default behavior?

No.

veprbl commented on src/algorithms/reco/JetReconstruction.cc:
```diff
@@ -23,31 +23,32 @@ namespace eicrecon {
   }
 
-  std::vector<edm4eic::ReconstructedParticle*> JetReconstruction::execute(
+  edm4eic::ReconstructedParticleCollection* JetReconstruction::execute(
```
```suggestion
  std::unique_ptr<edm4eic::ReconstructedParticleCollection> JetReconstruction::execute(
```

ruse-traveler replied:
Whoops! Thank you for catching that: I deleted the wrong line while cleaning up.

wdconinc commented on src/algorithms/reco/JetReconstruction.cc:
```diff
@@ -23,31 +23,32 @@ namespace eicrecon {
+
       // Only cluster particles within the given pt Range
       if ((mom->pt() > m_minCstPt) && (mom->pt() < m_maxCstPt)) {
-        particles.emplace_back(mom->px(), mom->py(), mom->pz(), mom->e() );
+        particles.push_back( PseudoJet(mom->px(), mom->py(), mom->pz(), mom->e()) );
```
You are reverting changes in the target branch.

ruse-traveler replied:
Ah! Thank you for catching that!

veprbl commented on src/algorithms/reco/JetReconstruction.cc:
```diff
@@ -57,41 +58,48 @@ namespace eicrecon {
+        cst_output.setMomentum(edm4hep::Vector3f(csts[j].px(), csts[j].py(), csts[j].pz()));
+        cst_output.setEnergy(csts[j].e());
+        cst_output.setMass(csts[j].m());
+        //jet_output.addToParticles(cst_output);  // FIXME: global issue with podio reference
```
This one would need to be routed properly, something along the lines of
```c++
add_to_particle(jet_output, input_collection[csts[j].user_index()])
```
where
```c++
template <typename T>
add_to_particle(edm4eic::MutableReconstructedParticle &p, T &other);

template <>
add_to_particle(edm4eic::MutableReconstructedParticle &p, edm4eic::ReconstructedParticle &other) {
  p.addToParticles(other);
}

template <>
add_to_particle(edm4eic::MutableReconstructedParticle &p, edm4eic::MCParticle &other) {
  p.addToMCParticles(other);
}
```

veprbl commented on src/algorithms/reco/JetReconstruction.h:
```diff
@@ -6,11 +6,13 @@
 // standard c includes
 #include <vector>
-// DD4hep includes
+// jana includes
```
```suggestion
```

veprbl commented on src/algorithms/reco/JetReconstruction.cc:
```diff
@@ -23,31 +23,32 @@ namespace eicrecon {
 
-  std::vector<edm4eic::ReconstructedParticle*> JetReconstruction::execute(
+  edm4eic::ReconstructedParticleCollection* JetReconstruction::execute(
     const std::vector<const edm4hep::LorentzVectorE*> momenta) {
```
Would need to use templated type here
```suggestion
    const T &input_collection) {
```

veprbl reviewed (commented):
Looks good. Mostly rambling about unusual formatting.
Please use smart pointers. We also would like to get proper associations, I'm putting down some ideas on how to integrate that.

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -37,7 +36,6 @@ void InitPlugin(JApplication *app) {
     app->Add(new JChainMultifactoryGeneratorT<MatchClusters_factory>(
         "ReconstructedParticlesWithAssoc",
         { "EcalEndcapNClusters",
-          "EcalBarrelScFiClusters",
```
This doesn't seem right...

ruse-traveler replied:
😩 Aggh. I deleted that by accident. Thank you for catching that!

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -65,22 +63,17 @@ void InitPlugin(JApplication *app) {
+    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
+             "GeneratedJets",
+            {"MCParticles"},
+            {"GeneratedJets"},
```
We don't need to change these factories to mutifactories at all. They need to output jets (currently modeled as `edm4hep::ReconstructedParticles`, we then refer to constituents that are already written in the MCParticles and ReconstructedParticles collections.

ruse-traveler replied:
Out of curiosity: if we wanted to have multiple jet collections (e.g. a collection of track-only jets and then a collection of particle flow jets), would it be ideal to use multifactories?

ruse-traveler replied:
Ahhhh I see. I guess for my edification: are multifactories exclusively for collections that require multiple inputs (e.g. reconstructed particles)?

veprbl replied:
Multifactories are needed for factories that want to produce multiple outputs.

Just to clarify (cc @wdconinc), we are likely to convert all factories (even those producing a single output) to multfactories eventually. Presumably that would not impact the performance, but a consistent interface would simplify implementation of Gaudi-style steering via configuration in Python. Latter is the current informal consensus regarding the direction to go forward. Having said that, producing multiple kinds of jets from the same factories doesn't seem like a sound choice: there is no shared computation between the two procedures, but the shared state opens them up to introducing bugs. We would be much better off making algorithms sufficiently configurable and running those from separately defined factories. In that sense, you might be doing the right thing for the wrong reason.

bspage912 commented on src/global/reco/reco.cc:
```diff
@@ -65,22 +63,17 @@ void InitPlugin(JApplication *app) {
+    ));
+    app->Add(new JChainMultifactoryGeneratorT<ReconstructedJets_factory>(
+            "ReconstrucctedJets",
+            {"ReconstructedParticles"},
```
Change input collection to ReconstructedChargedParticles?

bspage912 commented on src/algorithms/reco/JetReconstruction.cc:
```diff
@@ -23,31 +23,32 @@ namespace eicrecon {
     for (const auto &mom : momenta) {
+
       // Only cluster particles within the given pt Range
       if ((mom->pt() > m_minCstPt) && (mom->pt() < m_maxCstPt)) {
```
May be for another PR to address, but logically, I'm not sure we want to impose the particle / track pT cuts here as we likely want them to be different depending on the input. We usually place some threshold on tracks (and calorimeter clusters) based on the tracker performance, but place a much lower pT threshold on stable MC particles.

ruse-traveler replied:
Definitely! I was thinking on addressing this in another PR, but in making the options for both factories configurable, we'll also be able to set the kinematic thresholds differently between the reconstructed and generated jets.

bspage912 reviewed (commented):
I have no comment on the code changes themselves as I am still trying to understand all this myself :-). 

I added comments on using ReconstructedChargedParticles vs ReconstructedParticles as the input in the ReconstructedJets_factory, on possibly restricting the input particle species in GeneratedJets_factory, and on where we logically want to place input track/particle pT thresholds.

ruse-traveler commented on src/global/reco/ReconstructedJets_factory.cc:
```diff
@@ -1,49 +1,74 @@
+
 
     void ReconstructedJets_factory::Process(const std::shared_ptr<const JEvent> &event) {
-        auto rc_particles = event->Get<edm4eic::ReconstructedParticle>("ReconstructedParticles");
```
Whoops! Good catch: that should be ReconstructedChargedParticles. But that's correct! For multifactories, the input and output tags are specified in the reco.cc file.

ruse-traveler commented on src/global/reco/GeneratedJets_factory.cc:
```diff
@@ -1,52 +1,77 @@
+
+            const auto& momentum = particle -> getMomentum();
+            const auto& energy   = particle -> getEnergy();
+            momenta.push_back(new edm4hep::LorentzVectorE(momentum.x, momentum.y, momentum.z, energy));
```
That's a good point... And I would say we should. But what if we add an option to the algorithm to pick either charged or all generated particles?

ruse-traveler replied:
Forgot to comment here: but to implement this, I added an additional condition to the generated jets factory to pick out only charged particles!

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -65,22 +63,17 @@ void InitPlugin(JApplication *app) {
+            app
+    ));
+    app->Add(new JChainMultifactoryGeneratorT<ReconstructedJets_factory>(
+            "ReconstrucctedJets",
```
```suggestion
            "ReconstructedJets",
```

veprbl commented on src/global/reco/ReconstructedJets_factory.h:
```diff
@@ -3,37 +3,49 @@
 #pragma once
 
-#include <edm4eic/ReconstructedParticleCollection.h>
+// jana includes
```
```suggestion
```

ruse-traveler commented:
Hi all, just wanted to touch base on this PR to make sure it gets in before the September campaign. If everyone's okay with implementing PODIO references in a separate PR, then I think I've addressed all comments so far. (If no one objects, then I'll mark all the above conversations as resolved.)

Let me know if I missed anything or if there's anything else I need to do to make sure this gets merged!

wdconinc commented:
> Let me know if I missed anything or if there's anything else I need to do to make sure this gets merged!

There are conflicts...

ruse-traveler commented:
> > Let me know if I missed anything or if there's anything else I need to do to make sure this gets merged!
> 
> There are conflicts...

Whoops! I should've specified: aside from the conflicts! 😓 

(I'll take care of those, though!)

veprbl commented on src/global/reco/GeneratedJets_factory.cc:
```diff
@@ -1,52 +1,72 @@
+
+        // loop over input tags
+        size_t iInput = 0;
+        for (const std::string& input_tag : GetInputTags()) {
```
Why are you iterating over input collections? This can't work because down below you overwrite previous result with the latest collection.

I would stick to just one input particle collection `GetInputTags()[0]` and leave the other inputs reserved for possible future use. If we need to merge multiple particle collections, let's make that code into a separate factory.

ruse-traveler replied:
Ah! This is a holdover from my misunderstanding of what a multifactory was for: my thinking was that I would loop over specified input collections and produce an output for each one. I didn't remove it in the last batch of commits just because I thought it wouldn't make a difference.

But I agree with your suggestion: I'll stick to `GetTags()[0]` and reserve the others for future use!

veprbl commented on src/algorithms/reco/JetReconstruction.cc:
```diff
@@ -23,26 +23,27 @@ namespace eicrecon {
       // Only cluster particles within the given pt Range
       if ((mom->pt() > m_minCstPt) && (mom->pt() < m_maxCstPt)) {
-        particles.emplace_back(mom->px(), mom->py(), mom->pz(), mom->e() );
+        particles.emplace_back( PseudoJet(mom->px(), mom->py(), mom->pz(), mom->e()) );
```
```suggestion
        particles.emplace_back(mom->px(), mom->py(), mom->pz(), mom->e());
```

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -34,7 +33,7 @@ void InitPlugin(JApplication *app) {
         "ReconstructedParticlesWithAssoc",
         { "EcalEndcapNClusters",
-          "EcalBarrelScFiClusters",
+          "EcalBarrelSciFiClusters",
```
```suggestion
          "EcalBarrelScFiClusters",
```

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -62,22 +61,17 @@ void InitPlugin(JApplication *app) {
-        "EcalLumiSpecClusterAssociations",
-        },
-        "ReconstructedElectrons"
+    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
```
Submit a separate PR for this
```suggestion
    app->Add(new JChainFactoryGeneratorT<ReconstructedElectrons_factory>(
         {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
         "EcalBarrelScFiClusterAssociations",
         "EcalEndcapNClusterAssociations",
         "EcalEndcapPClusterAssociations",
         "EcalEndcapPInsertClusterAssociations",
         "EcalLumiSpecClusterAssociations",
         },
         "ReconstructedElectrons"

    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
```

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -19,7 +19,6 @@
 #include "GeneratedJets_factory.h"
 #include "ReconstructedJets_factory.h"
-#include "ReconstructedElectrons_factory.h"
 
```
```suggestion
#include "ReconstructedElectrons_factory.h"

```

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -62,22 +62,27 @@ void InitPlugin(JApplication *app) {
+         "EcalLumiSpecClusterAssociations",
+         },
+         "ReconstructedElectrons"
+
```
```suggestion
    ));

```

veprbl commented on src/global/reco/GeneratedJets_factory.h:
```diff
@@ -3,37 +3,47 @@
-        }
+
+        // ctor
+        explicit GeneratedJets_factory(std::string tag,
+                                       const std::vector<std::string>& input_tags,
```
```suggestion
        explicit GeneratedJets_factory(const std::vector<std::string>& input_tags,
```

veprbl commented on src/global/reco/ReconstructedJets_factory.h:
```diff
@@ -3,37 +3,47 @@
-        }
+
+        // ctor
+        explicit ReconstructedJets_factory(std::string tag,
+                                           const std::vector<std::string>& input_tags,
```
```suggestion
        explicit ReconstructedJets_factory(const std::vector<std::string>& input_tags,
```

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -62,22 +62,28 @@ void InitPlugin(JApplication *app) {
     ));
 
+    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
+             "GeneratedJets",
```
```suggestion
```

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -62,22 +62,28 @@ void InitPlugin(JApplication *app) {
+            app
+    ));
+    app->Add(new JChainMultifactoryGeneratorT<ReconstructedJets_factory>(
+            "ReconstructedJets",
```
```suggestion
```

veprbl commented on src/global/reco/ReconstructedJets_factory.cc:
```diff
@@ -1,49 +1,63 @@
-        auto rc_particles = event->Get<edm4eic::ReconstructedParticle>("ReconstructedParticles");
 
+        // grab input collection
+        auto input = event->Get<edm4eic::ReconstructedParticle>(GetInputTags()[0]);
```
Avoid conversion to `std::vector`
```suggestion
        auto input = static_cast<const edm4eic::ReconstructedParticleCollection*>(event->GetCollectionBase(GetInputTags()[0]));
```

veprbl commented on src/global/reco/GeneratedJets_factory.cc:
```diff
@@ -1,52 +1,67 @@
-        auto mc_particles = event->Get<edm4hep::MCParticle>("MCParticles");
 
+        // grab input collection
+        auto input = event->Get<edm4hep::MCParticle>(GetInputTags()[0]);
```
Avoid conversion to `std::vector`
```suggestion
        auto input = static_cast<const edm4hep::MCParticleCollection*>(event->GetCollectionBase(GetInputTags()[0]));
```

veprbl commented on src/global/reco/GeneratedJets_factory.h:
```diff
@@ -3,37 +3,47 @@
-        JetReconstruction m_jet_algo;
 
-    };
+        std::vector<std::string> m_input_tags;
```
Avoid adding dead code
```suggestion
```

veprbl commented on src/global/reco/ReconstructedJets_factory.h:
```diff
@@ -3,37 +3,47 @@
-        JetReconstruction m_jet_algo;
 
-    };
+        std::vector<std::string> m_input_tags;
```
```suggestion
```

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -62,22 +62,28 @@ void InitPlugin(JApplication *app) {
-        "EcalLumiSpecClusterAssociations",
-        },
-        "ReconstructedElectrons"
+         {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
+         "EcalBarrelScFiClusterAssociations",
+         "EcalEndcapNClusterAssociations",
+         "EcalEndcapPClusterAssociations",
+         "EcalEndcapPInsertClusterAssociations",
+         "EcalLumiSpecClusterAssociations",
+         },
+         "ReconstructedElectrons"
```
```suggestion
        {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
        "EcalBarrelScFiClusterAssociations",
        "EcalEndcapNClusterAssociations",
        "EcalEndcapPClusterAssociations",
        "EcalEndcapPInsertClusterAssociations",
        "EcalLumiSpecClusterAssociations",
        },
        "ReconstructedElectrons"
```

veprbl commented on src/global/reco/GeneratedJets_factory.cc:
```diff
@@ -1,52 +1,67 @@
-          delete mom;
+            // select only final state charged particles
+            const bool is_final_state = (particle->getGeneratorStatus() == 1);
+            const bool is_charged = (particle->getCharge() != 0.);
+            if (!is_final_state || !is_charged) continue;
```
Doesn't fit the description - please submit separately
```suggestion
            if (!is_final_state) continue;
```

veprbl commented on src/global/reco/reco.cc:
```diff
@@ -62,22 +62,28 @@ void InitPlugin(JApplication *app) {
+    ));
+    app->Add(new JChainMultifactoryGeneratorT<ReconstructedJets_factory>(
+            "ReconstructedJets",
+            {"ReconstructedChargedParticles"},
```
Please submit separately
```suggestion
            {"ReconstructedParticles"},
```

veprbl commented on src/global/reco/GeneratedJets_factory.cc:
```diff
@@ -1,52 +1,66 @@
-
+#include <services/log/Log_service.h>
+// event data model definitions
 #include <edm4hep/MCParticle.h>
```
```suggestion
#include <edm4hep/MCParticleCollection.h>
```

veprbl commented on src/global/reco/ReconstructedJets_factory.cc:
```diff
@@ -1,49 +1,63 @@
-#include "services/log/Log_service.h"
+#include <services/log/Log_service.h>
+// event data model definitions
+#include <edm4eic/ReconstructedParticle.h>
```
```suggestion
#include <edm4eic/ReconstructedParticleCollection.h>
```

veprbl commented on src/global/reco/ReconstructedJets_factory.cc:
```diff
@@ -1,49 +1,63 @@
+        for (const auto& particle : input) {
+
+            // TODO: Need to exclude the scattered electron
+            const auto& momentum = particle->getMomentum();
+            const auto& energy = particle->getEnergy();
```
```suggestion
            const auto& momentum = particle.getMomentum();
            const auto& energy = particle.getEnergy();
```

veprbl commented on src/global/reco/GeneratedJets_factory.cc:
```diff
@@ -1,52 +1,66 @@
+            const bool is_final_state = (particle->getGeneratorStatus() == 1);
+            if (!is_final_state) continue;
+
+            const auto& momentum = particle->getMomentum();
+            const auto& energy = particle->getEnergy();
```
```suggestion
            const auto& momentum = particle.getMomentum();
            const auto& energy = particle.getEnergy();
```

veprbl commented on src/global/reco/GeneratedJets_factory.cc:
```diff
@@ -1,52 +1,66 @@
-        for (const auto &mom : momenta) {
-          delete mom;
+            // select only final state charged particles
+            const bool is_final_state = (particle->getGeneratorStatus() == 1);
```
```suggestion
            const bool is_final_state = (particle.getGeneratorStatus() == 1);
```

veprbl commented:
I took liberty to finish the transitions from vectors to collections, fixed some issues introduced while rebasing and reverted charged particle cut.

veprbl reviewed (approved):
Checked outputs before/after. This PR introduces only minimal changes: jets are placed into collections before the constituents.

ruse-traveler commented:
@veprbl Thank you for taking care of those! Out of curiosity: is there a reason to revert the charged particle cut?

Either way, it's something we can discuss more when we revisit the constituent PODIO relations (i.e. for the next campaign). So this all looks good to me, and I'm okay with merging if everyone else is!

veprbl commented:
Bundling changes places unnecessary burden on reviewers. Switch to multifactories should not cause any changes in the output, and it should be easy to demonstrate that. Thus, changes to the cuts don't belong in this change.

veprbl commented:
@ruse-traveler Can you resubmit charge cut for review, or do you want me to do it? Please, create new branch in this repo this time. I also suggest renaming the collection to {Reconstructed,Generated}ChargedJets to communicate the difference in the result.

ruse-traveler commented:
> @ruse-traveler Can you resubmit charge cut for review, or do you want me to do it? Please, create new branch in this repo this time. I also suggest renaming the collection to {Reconstructed,Generated}ChargedJets to communicate the difference in the result.

Can do! I'll put together a PR for that using a branch from this repo. And makes sense (re: reverting the cut)! I'll make sure to not sneak in extra changes in future PRs...
